### ISSUE TITLE
System.ArgumentNullException by System.Threading.Monitor.Enter

### ISSUE THREAD

**Issue by sebastianburckhardt** [dotnet/orleans#3352]
I am getting System.ArgumentNullException 
 "Value cannot be null."
   at System.Threading.Monitor.Enter(Object obj)
   at Orleans.Runtime.ActivationCollector.ScanAll(TimeSpan ageLimit)

the debugger shows this is thrown when calling ForceActivation, on line 321 of 
https://github.com/dotnet/orleans/blob/master/src/OrleansRuntime/Catalog/ActivationCollector.cs 

            return result ?? nothing;

It looks to me like the static definition of `nothing` may not be thread safe:

```csharp
private static readonly List<ActivationData> nothing = new List<ActivationData> { Capacity = 0 };
```

**Comment by sebastianburckhardt** [comment]
I think the debugger was not showing me the right line. The thing that is really throwing, I think, is this line earlier in the same method:

```csharp
lock (activation)
```

This is more consistent with the error message (cannot lock null).
And surely enough, the problem goes away if I add a null-check.

**Comment by sergeybykov** [comment]
@sebastianburckhardt So is there an issue or not?

**Comment by sebastianburckhardt** [comment]
Yes, there is an issue. The runtime throws a `System.ArgumentNullException` from https://github.com/dotnet/orleans/blob/master/src/OrleansRuntime/Catalog/ActivationCollector.cs
whenever I initiate a collection of all actors using the management grain.

The problem goes away if I add a null check around the offending section
```csharp
if (activation != null)
{
   lock(activation)
```
so that's how I have worked around the issue for now.

**Comment by benjaminpetit** [comment]
I managed to reproduce the issue.

It seems that in `ActivationCollector.ScanAll`, when we want to deactivate a grain activation, we set the ActivationData in the bucket to `null`, but we don't remove it from the bucket, so if you call `ForceActivationCollection` two times in a row you get the exception.

**Comment by gabikliot** [comment]
I think there was already a bug about that where I suggested to remove the code that sets to null and just remove. Need to find that bug.

**Comment by benjaminpetit** [comment]
Yes I think this issue is linked with #3165. Let's fix it once for all :)

**Comment by sergeybykov** [comment]
Resolved via #3399.
